heaviside/ast.py

```diff
@@ -247,9 +288,70 @@ class ASTStateTask(ASTState):
                        ASTModRetry,
                        ASTModCatch]
 
-    def __init__(self, state, arn, block):
-        super(ASTStateTask, self).__init__(state, block)
-        self.arn = arn
+    valid_services = ['Arn',
+                      'Lambda',
+                      'Activity']
+
+    def __init__(self, service, function, args_kwargs, block):
+        super(ASTStateTask, self).__init__(service, block)
+
+        if service.value not in self.valid_services and \
+           service.value not in AWS_SERVICES.keys():
+            service.raise_error('Invalid Task service')
+
+        if function is None:
+            if service.value in ('Lambda', 'Activity', 'Arn'):
+                if len(args_kwargs.args) == 0:
+                    service.raise_error('{} task requires a function name argument'.format(service.value))
+
+                function = args_kwargs.args.pop(0)
+            else:
+                service.raise_error('{} task requires a function to call'.format(service.value))
+        else:
+            if service.value in ('Lambda', 'Activity', 'Arn'):
+                function.raise_error('Unexpected function name')
+            else:
+                try:
+                    function.lookup = function.value # Save value for looking up when checking kwargs
+                    function.value = AWS_SERVICES[service.value][function.value]['name']
+                except KeyError:
+                    function.raise_error('Invalid Task function')
+
+        if service.value == 'Arn' and not function.value.startswith('arn:aws:'):
+            function.raise_error("ARN must start with 'arn:aws:'")
+        if len(args_kwargs.args) > 0:
+            args_kwargs.args[0].raise_error('Unexpected argument')
+        if service.value in ('Lambda', 'Activity') and len(args_kwargs.kwargs) > 0:
+            tuple(args_kwargs.kwargs.keys())[0].raise_error('Unexpected keyword argument')
+
+        if service.value not in ('Lambda', 'Activity', 'Arn'):
+            required = AWS_SERVICES[service.value][function.lookup]['required_keys']
```

Comment:
Do you need to make a copy of this since you're mutating it below?

Reply:
Good catch